src/snapred/backend/recipe/algorithm/FocusSpectraAlgorithm.py

```diff
@@ -0,0 +1,126 @@
+        if not self.mantidSnapper.mtd.doesExist(self.getPropertyValue("InputWorkspace")):
+            errors["InputWorkspace"] = "Input workspace does not exist"
+            return errors
+        if not self.mantidSnapper.mtd.doesExist(self.getPropertyValue("GroupingWorkspace")):
+            errors["GroupingWorkspace"] = "Grouping workspace does not exist"
+            return errors
+        if not self.getPropertyValue("Ingredients"):
+            errors["Ingredients"] = "Ingredients are required"
+            return errors
+
+        inWS = self.mantidSnapper.mtd[self.getPropertyValue("InputWorkspace")]
+        groupWS = self.mantidSnapper.mtd[self.getPropertyValue("GroupingWorkspace")]
+        if "Grouping" not in groupWS.id():
+            errors["GroupingWorkspace"] = "Grouping workspace must be an actual GroupingWorkspace"
+        elif inWS.getNumberHistograms() == len(groupWS.getGroupIDs()):
+            msg = "The data appears to have already been diffraction focused"
+            errors["InputWorkspace"] = msg
+            errors["GroupingWorkspace"] = msg
+        elif inWS.getNumberHistograms() != groupWS.getNumberHistograms():
+            msg = f"""
+                The workspaces {self.getPropertyValue("InputWorkspace")}
+                and {self.getPropertyValue("GroupingWorkspace")}
+                have inconsistent number of spectra
+                """
+            errors["InputWorkspace"] = msg
+            errors["GroupingpWorkspace"] = msg
+        return errors
+
+    def PyExec(self):
+        self.ingredients: Ingredients = Ingredients.parse_raw(self.getProperty("Ingredients").value)
+        self.chopIngredients(self.ingredients)
+        self.unbagGroceries()
+
+        self.log().notice("Execute of FocusSpectra START!")
+
+        # convert to d-spacing and diffraction focus and rebin ragged
+        self.mantidSnapper.ConvertUnits(
+            "Converting to Units of dSpacing...",
+            InputWorkspace=self.inputWSName,
+            Emode="Elastic",
+            Target="dSpacing",
+            OutputWorkspace=self.outputWSName,
+            ConvertFromPointData=True,
+        )
+        self.mantidSnapper.DiffractionFocussing(
+            "Performing Diffraction Focusing ...",
+            InputWorkspace=self.outputWSName,
+            GroupingWorkspace=self.groupingWSName,
+            OutputWorkspace=self.outputWSName,
+            PreserveEvents=True,
+        )
+        self.mantidSnapper.RebinRagged(
+            "Rebinning ragged bins...",
+            InputWorkspace=self.outputWSName,
+            XMin=self.dMin,
+            XMax=self.dMax,
+            Delta=self.dBin,
+            OutputWorkspace=self.outputWSName + "_noevents",
+            PreserveEvents=False,
+        )
+        self.mantidSnapper.DeleteWorkspace("Delete intermediate ws", Workspace=self.outputWSName)
+        self.setProperty("OutputWorkspace", self.outputWSName + "_noevents")
```

Comment:
I'm not sure what these two lines are doing.  I don't think we want to change the outputworkspace from the user specification.

Reply:
this has to do with that one issue we had with mantid where if we didnt explicitly state the outputworkspace with a setproperty it wouldnt persist.  I'll see if I can refactor this to avoid this hack